src/algorithm/convexhull.rs

```diff
@@ -0,0 +1,199 @@
+use std::collections::BTreeSet;
+
+// Determine whether a point lies on one side of a line segment, or the other.
+// The cross product v x w of two vectors v and w is a vector whose length is
+// |v||w|sin φ, (where |v| is the length of v and φ is the angle between the vectors),
+// and which is orthogonal (perpendicular) to both v and w.  Since there are two
+// such possible vectors, the definition arbitrarily selects the one that matches
+// the direction in which a screw would move if rotated from v to w
+
+// Mathematically, if the coordinates of vectors v and w are (vx, vy) and (wx, wy)
+// respectively, the cross product will be (vxwy - vywx). If a segment is
+// defined by points A B and, we wish to check on which side of AB a third point C falls,
+// we can compute the cross product AB x AC and check its sign:
+// If it's negative, it will be on the "right" side of AB
+// (when standing on A and looking towards B). If positive, it will be on the left side
+fn point_location<T>(p_a: &Point<T>, p_b: &Point<T>, p_c: &Point<T>) -> bool
+    where T: Float
+{
+    (p_b.x() - p_a.x()) * (p_c.y() - p_a.y()) - (p_b.y() - p_a.y()) * (p_c.x() - p_a.x()) >
+    T::zero()
+}
+
+// Fast distance between line segment (p_a, p_b), and point p_c
+fn pseudo_distance<T>(p_a: &Point<T>, p_b: &Point<T>, p_c: &Point<T>) -> T
+    where T: Float
+{
+    let abx = p_b.x() - p_a.x();
+    let aby = p_b.y() - p_a.y();
+    let dist = abx * (p_a.y() - p_c.y()) - aby * (p_a.x() - p_c.x());
+    if dist < T::zero() { -dist } else { dist }
+}
+
+// Adapted from http://www.ahristov.com/tutorial/geometry-games/convex-hull.html
+fn quick_hull<T>(points: &[Point<T>]) -> Vec<Point<T>>
+    where T: Float
+{
+    // can't build a hull from fewer than four points
+    if points.len() < 4 {
+        return points.to_vec();
+    }
+    let mut hull: Vec<Point<T>> = vec![];
+    let mut min_x_idx = <usize>::min_value();
+    let mut max_x_idx = <usize>::min_value();
+    let mut min_x = Float::max_value();
+    let mut max_x = Float::min_value();
+    let to_retain: BTreeSet<usize> = (0..points.len() - 1).collect();
+    let mut to_remove: BTreeSet<usize> = BTreeSet::new();
+    for (idx, point) in points.iter().enumerate() {
+        if point.x() < min_x {
+            min_x = point.x();
+            min_x_idx = idx;
+        }
+        if point.x() > max_x {
+            max_x = point.x();
+            max_x_idx = idx;
+        }
+    }
+    to_remove.insert(min_x_idx);
+    to_remove.insert(max_x_idx);
+
+    let p_a = points[min_x_idx];
+    let p_b = points[max_x_idx];
+    // min x and max x points are always part of the hull
+    hull.push(p_a);
+    hull.push(p_b);
+    let mut left_set: Vec<Point<T>> = vec![];
+    let mut right_set: Vec<Point<T>> = vec![];
+    // divide remaining points into left and right
+    // this is a bit hairy
+    for point in &to_retain.difference(&to_remove)
+        .map(|&idx| points[idx])
+        .collect::<Vec<Point<T>>>() {
```

Comment:
Removing the `collect()` gives me a trait bound error:

`the trait std::iter::Iterator is not implemented for &std::iter::Map<std::collections::btree_set::Difference<'_, usize>, [closure@convexhull.rs:73:14: 73:32 points:_]>`